dev-dotnet/dotnet-sdk/dotnet-sdk-5.0.203.ebuild

```
@@ -0,0 +1,238 @@
+	#  but the reference-packages have just been installed in /usr/lib
+	for x in "${DOTNET_ROOT}"/reference-packages "${BROOT}"/usr/lib/dotnet/reference-packages; do
+		if [[ -d "${x}" ]]; then
+			DOTNET_REFERENCE="${x}"
+			break
+		fi
+	done
+
+	if [[ ! -d "${DOTNET_REFERENCE}" ]]; then
+		die "Can't find installed .NET Source-build Reference Packages"
+	fi
+}
+
+src_unpack() {
+	unpack "${MY_P}.tar.gz"
+	dotnet_unpack
+}
+
+src_prepare() {
+	# add_patch <submodule> <file.patch>
+	add_patch() {
+		cp "${2}" "patches/${1}/9000-$(basename "${2}")" || die
+	}
+
+	eapply "${FILESDIR}"/dotnet-sdk-5.0.203-additional-arguments.patch
+	add_patch sdk "${DISTDIR}"/dotnet-sdk-telemetry-optout.patch
+	add_patch runtime "${FILESDIR}"/dotnet-sdk-5.0.203-runtime-add-gentoo-ids.patch
+	add_patch runtime "${FILESDIR}"/dotnet-sdk-5.0.203-runtime-additional-arguments.patch
+	add_patch runtime "${FILESDIR}"/dotnet-sdk-5.0.203-runtime-as-needed.patch
+	add_patch runtime "${FILESDIR}"/dotnet-sdk-5.0.203-runtime-locate-ranlib.patch
+
+	default
+}
+
+src_configure() {
+	dotnet_find
+
+	# SDK is modified in-place...
+	if use system-bootstrap; then
+		mkdir -p dotnet || die
+		ln -s "${DOTNET_ROOT}"/{host,packs,shared} dotnet || die
+		cp -a "${DOTNET_ROOT}"/{dotnet,sdk} dotnet || die
+	else
+		ln -s "${DOTNET_ROOT}" dotnet || die
+	fi
+}
+
+src_compile() {
+	dotnet_find
+	dotnet_find_reference
+
+	# Required for --with-packages to work
+	rm -rf packages/archive
+
+	local mybuildargs=(
+		--with-sdk dotnet
+		--with-packages "${DOTNET_ROOT}"/source-artifacts
+		--with-ref-packages "${DOTNET_REFERENCE}"
+		--
+
+		/p:UseSystemLibraries=true
+		/p:UseSystemLibunwind=$(usex system-libunwind true false)
+
+		/p:LogVerbosity=normal
+		/p:MinimalConsoleLogOutput=false
+		/verbosity:normal
+
+		# Not sure if this has any effect on the subprocesses
+		/maxCpuCount:$(makeopts_jobs)
+
+		# /p:AdditionalBuildArguments are passed to src/runtime.*/eng/build.sh
+		# /p:NativeBuildArguments are passed to src/runtime.*/eng/native/build-commons.sh
+		/p:AdditionalBuildArguments="-cmakeargs -DCMAKE_VERBOSE_MAKEFILE=ON"
+		/p:NativeBuildArguments="${CMAKE_GENERATOR_FLAG} -numproc $(makeopts_jobs)"
+		# If any dotnet maintainer is reading this: Please honor the
+		#  CMAKE_MAKEFILE_GENERATOR and MAKEFLAGS variables if set.
+	)
+
+	# Adapted from www-client/firefox
+	if ${DOTNET_FORCE_CLANG} && ! tc-is-clang; then
+		einfo "Enforcing the use of Clang due to build errors..."
+		AR=llvm-ar
+		CC=${CHOST}-clang
+		CXX=${CHOST}-clang++
+		NM=llvm-nm
+		OBJCOPY=llvm-objcopy
+		OBJDUMP=llvm-objdump
+		RANLIB=llvm-ranlib
+
+		if is-flagq '-flto*' && ! has_version 'sys-devel/clang[default-lld]' && ! is-ldflagq '-fuse-ld=lld'; then
+			has_version sys-devel/lld || die "Please install sys-devel/lld (or sys-devel/clang[default-lld]) to build with LTO"
+			einfo "Enforcing the use of LLD for Clang LTO..."
+			append-ldflags '-fuse-ld=lld'
+		fi
+
+		strip-unsupported-flags
+	fi
```

```suggestion
	if ${DOTNET_FORCE_CLANG}; then
		if ! tc-is-clang; then
			einfo "Enforcing the use of Clang due to build errors..."
			AR=llvm-ar
			CC=${CHOST}-clang
			CXX=${CHOST}-clang++
			NM=llvm-nm
			OBJCOPY=llvm-objcopy
			OBJDUMP=llvm-objdump
			RANLIB=llvm-ranlib
			strip-unsupported-flags
		fi

		if is-flagq '-flto*' && ! has_version 'sys-devel/clang[default-lld]' && ! is-ldflagq '-fuse-ld=lld'; then
			has_version sys-devel/lld || die "Please install sys-devel/lld (or sys-devel/clang[default-lld]) to build with LTO"
			einfo "Enforcing the use of LLD for Clang LTO..."
			append-ldflags '-fuse-ld=lld'
		fi

		# Build failures with libcxx
		if has_version 'sys-devel/clang[default-libcxx]' || is-flagq '-stdlib=libc++'; then
			append-cxxflags '-stdlib=libstdc++'
			append-ldflags '-stdlib=libstdc++'
		fi
	fi
```

The LTO & libcxx workarounds should be applied regardless of `tc-is-clang`, if DOTNET_FORCE_CLANG is true. I assume DOTNET_FORCE_CLANG will eventually be replaced by a `clang` use flag? Perhaps it could be replaced now and just forced in profile?

---

The `clang` and `lto` USE flags seem very inconsistent in the tree right now, and since we only want to control whether we *force* clang or not (probably on a version-by-version basis), I think this is the saner approach. Unlike firefox (where using clang allows LTO to work across the rust/c++ boundary, and LTO is handled by the build system instead of `*FLAGS`), there's no inherent benefit to using clang when GCC also builds correctly, so I doubt users would want to toggle it. I'm open to changing it, but if you phrase it like a question, my answer is I wouldn't.

I was mostly assuming that if the user has set up clang as their default compiler *and* enabled LTO, they would likely be using LLD already, as not doing that breaks many other builds or just doesn't LTO at all. This check is purely for GCC users running full-system LTO setups that might wish to install LLD to also LTO this package. That said, `-flto=n` will get stripped by `strip-unsupported-flags`, so this might not be extremely effective unless the user is using `-flto` raw.
That said, the libstdc++ check is different, I think I'll leave that outside of the DOTNET_FORCE_CLANG variable, instead only apply it if `tc-is-clang`, as this should be applied regardless of whether the ebuild forces clang or not.